---
title: "RNAseq QC"
author: "Sarah Kay Buddenborg"
date: "`r Sys.Date()`"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prepare environment

```{r environment, warning = FALSE, message = FALSE, results = FALSE}
if (!requireNamespace('BiocManager', quietly = TRUE))
  install.packages('BiocManager')

BiocManager::install("tidyverse", force = TRUE)
BiocManager::install("ggplot2", force = TRUE)
BiocManager::install("viridis", force = TRUE)
BiocManager::install("pheatmap", force = TRUE)
BiocManager::install("RColorBrewer", force = TRUE)
BiocManager::install("dplyr", force = TRUE)
BiocManager::install("gplots", force = TRUE)
BiocManager::install("dendsort", force = TRUE)
BiocManager::install("ggeasy", force = TRUE)
BiocManager::install("tidyr", force = TRUE)

library(tidyverse)
library(ggplot2)
library(viridis)
library(pheatmap)
library(RColorBrewer)
library(dplyr)
library(gplots)
library(dendsort)
library(ggeasy)
library(tidyr)
```

## PCA plot of all samples from STAR mapped reads quantified to TPMs using StringTie

Following
[https://tavareshugo.github.io/data-carpentry-rnaseq/00_exercises.html#31_examine_prcomp()](https://tavareshugo.github.io/data-carpentry-rnaseq/00_exercises.html#31_examine_prcomp())

```{r}
# Load the count matrix
TPM <- read_tsv("~/Desktop/STAR_stringtie/TPM_matrix.txt") %>%
# Add one to all TPM values to avoid #NA values
  mutate(across(where(is.numeric),~ .x + 1)) %>%
# Log-transform the TPM+1 values
  mutate(across(where(is.numeric),log2)) %>%
# Transpose the counts table so the "gene" column becomes the rownames
  column_to_rownames("gene") %>%
# Coerce into a matrix
  as.matrix() %>%
# Transpose the matrix so that rows = samples and columns = variables
  t()
```

Double check for #N/A and non-numeric values:

```{r}
all(is.finite(TPM))
```

Perform the PCA:

```{r}
sample_pca <- prcomp(TPM)

# Check output of prcomp:
TPM[1:10, 1:5]
```

Class of the object:

```{r}
class(sample_pca)
```

Structure of the object:

```{r}
str(sample_pca)
```

"sdev" contains the standard deviation explained by each PC, so if we
square it we get the eigenvalues (or explained variance) "rotation"
contains the variable loadings for each PC, which define the
eigenvectors "x" contains the PC scores, i.e. the data projected on the
new PC axis "center" in this case contains the mean of each gene, which
was subtracted from each value "scale" contains the value FALSE because
we did not scale the data by the standard deviation

eigenvalues squared:

```{r}
pc_eigenvalues <- sample_pca$sdev^2
```

PC scores:

```{r}
pc_scores <- sample_pca$x
```

Variable loadings:

```{r}
pc_loadings <- sample_pca$rotation
ncol(pc_scores)
length(pc_eigenvalues)
ncol(pc_loadings)
```

Convert matrix to tibble object in order to make a plot in ggplot2: Add
a new column with the percent variance Add another column with the
cumulative variance explained

```{r}
# Create a "tibble" manually with a variable indicating the PC number and a variable with the variances
pc_eigenvalues <- tibble(PC = factor(1:length(pc_eigenvalues)), 
                         variance = pc_eigenvalues) %>% 
  mutate(pct = variance/sum(variance)*100) %>% 
  mutate(pct_cum = cumsum(pct))
```

Print the result:

```{r}
pc_eigenvalues
```

Produce a Scree plot to show the fraction of the total variance
explained by each PC

```{r}
scree_plot <- pc_eigenvalues %>% 
  ggplot(aes(x = PC)) +
  geom_col(aes(y = pct)) +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  theme(axis.text.x = element_text(size = 5)) +
  labs(x = "Principal component", y = "Fraction variance explained")

scree_plot
```

Save Scree plot:

```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/scree.tiff", height = 7, width = 7, units = "in", device = 'tiff', dpi = 700)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/scree.png", height = 7, width = 7, units = "in", device = 'png', dpi = 700)
```

The PC scores are stored in the "x" value of the prcomp object:

```{r}
pc_scores <- sample_pca$x
```

Convert to a tibble retaining the sample names as a new column:

```{r}
pc_scores <- pc_scores %>% 
  as_tibble(rownames = "sample")
```

Print the result:

```{r}
pc_scores
```

Load sample info i.e. metadata:

```{r}
sample_info <- read_tsv("~/Desktop/STAR_stringtie/sample_info.txt")
```

Extract the loadings from prcomp Convert to a tibble retaining the
sample names as a new column Join with "sample_info" table Create the
plot

```{r}
# Get HEX codes for palette used in heatmaps so sample colors are same in all plots
stagecol <- c("#fde725","#a0da39","#4ac16d","#1fa187","#277f8e","#365c8d","#46327e","#440154")

pca_plot <- sample_pca$x %>% 
  as_tibble(rownames = "sample") %>% 
  full_join(sample_info, by = "sample") %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(colour = stage,shape = sex,alpha = 20),size = 5) +
  scale_colour_manual(values = stagecol,breaks = c("Eggs","Miracidia","1d_Sporocysts","5d_Sporocysts","32d_Sporocysts","Cercariae","2d_Somules","Juveniles")) +
  labs(y = "PC2 (23.7%)",x = "PC1 (36.6%)",title = "Principal Component ANALYSIS \n All Genes") +
  theme (panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         panel.background = element_blank(),
         legend.key = element_blank(),
         axis.line = element_line(colour = "gray")) +
  ggeasy::easy_center_title() +
  scale_alpha(guide = 'none')

pca_plot
```

Save the PCA plot:

```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/PCA.tiff", height = 7, width = 7, units = "in", device = 'tiff', dpi = 600)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/PCA.png", height = 7, width = 7, units = "in", device = 'png', dpi = 600)
```

Which genes have the most influence on each PC axis:

```{r}
pc_loadings <- sample_pca$rotation
pc_loadings <- pc_loadings %>% 
  as_tibble(rownames = "gene")

pc_loadings
```

What are the top 10 genes with highest loading on PC1 and PC2:

```{r}
slice <- dplyr::slice
top_genes <- pc_loadings %>% 
  # select only the PCs we are interested in
  select(gene, PC1, PC2) %>%
  # convert to a "long" format
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  # for each PC
  group_by(PC) %>% 
  # arrange by descending order of loading
  arrange(desc(abs(loading))) %>% 
  # take the 10 top rows
  slice(1:10) %>% 
  # pull the gene column as a vector
  pull(gene)

top_loadings <- pc_loadings %>% 
  filter(gene %in% top_genes)
```

Visualise variable loadings:

```{r}
loadings_plot <- ggplot(data = top_loadings) +
  geom_segment(aes(x = 0, y = 0, xend = PC1, yend = PC2), 
               arrow = arrow(length = unit(0.1, "in")),
               colour = "brown") +
  geom_text(aes(x = PC1, y = PC2, label = gene),
            nudge_y = 0.005, size = 3) +
  scale_x_continuous(expand = c(0.02, 0.02))

loadings_plot
```

```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/loadings.tiff", height = 7, width = 7, units = "in",device = 'tiff', dpi = 700)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/loadings.png", height = 7, width = 7, units = "in",device = 'png', dpi = 700)
```

## PCA of top 25% and 10% most highly variable genes

Following
<https://jdblischak.github.io/singlecell-qtl/pca-variable.html>

```{r}
TPMvar <- read_tsv("~/Desktop/STAR_stringtie/TPM_matrix.txt") %>%
# Add one to all TPM values to avoid #NA values
  mutate(across(where(is.numeric),~ .x + 1)) %>%
# Log-transform the TPM+1 values
  mutate(across(where(is.numeric),log2)) %>%
# Transpose the counts table so the "gene" column becomes the rownames
  column_to_rownames("gene") %>%
# Coerce into a matrix
  as.matrix()

all(is.finite(TPMvar))

# Calculate coefficient of variation
compute_cv <- function(x) sd(x) / mean(x)
cv <- apply(TPMvar, 1, compute_cv)
summary(cv)
```

# Select 25% of genes with highest CV

```{r}
cutoff25 <- 0.25
summary(cv[rank(cv) / length(cv) > 1 - cutoff25])
```

```{r}
variable25 <- TPMvar[rank(cv) / length(cv) > 1 - cutoff25, ]
dim(variable25)
```

# PCA using the 25% most variable genes

```{r}
var25_mat <- t(variable25)
var25_pca <- prcomp(var25_mat)
class(var25_pca)
str(var25_pca)
var25pc_eigenvalues <- var25_pca$sdev^2
var25pc_scores <- var25_pca$x
var25pc_loadings <- var25_pca$rotation
ncol(var25pc_scores)
length(var25pc_eigenvalues)
ncol(var25pc_loadings)
```

```{r}
var25pc_eigenvalues <- tibble(PC = factor(1:length(var25pc_eigenvalues)), 
                         variance25 = var25pc_eigenvalues) %>% 
  mutate(pct = variance25/sum(variance25)*100) %>% 
  mutate(pct_cum = cumsum(pct))

var25pc_eigenvalues
```

```{r}
var25scree_plot <- var25pc_eigenvalues %>% 
  ggplot(aes(x = PC)) +
  geom_col(aes(y = pct)) +
  geom_line(aes(y = pct_cum, group = 1)) + 
  geom_point(aes(y = pct_cum)) +
  theme(axis.text.x = element_text(size = 5)) +
  labs(x = "Principal component", y = "Fraction variance explained", cex = 0.5, title = "Principal Components \n Top 25% Variant Genes") +
  ggeasy::easy_center_title()

var25scree_plot
```

```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/var25scree.tiff", height = 7, width = 7, units = "in", device = 'tiff', dpi = 700)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/var25scree.png", height = 7, width = 7, units = "in", device = 'png', dpi = 700)
```

```{r}
var25pc_scores <- var25_pca$x
var25pc_scores <- var25pc_scores %>% 
  as_tibble(rownames = "sample")

var25pc_scores
```

```{r}
sample_info <- read_tsv("~/Desktop/STAR_stringtie/sample_info.txt")

#Get HEX codes for palette used in heatmaps so sample colors are same in all plots
stagecol <- c("#fde725","#a0da39","#4ac16d","#1fa187","#277f8e","#365c8d","#46327e","#440154")

var25pca_plot <- var25_pca$x %>% 
  as_tibble(rownames = "sample") %>% 
  full_join(sample_info, by = "sample") %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(colour = stage,shape = sex,alpha = 20),size = 5) +
  scale_colour_manual(values = stagecol,breaks = c("Eggs","Miracidia","1d_Sporocysts","5d_Sporocysts","32d_Sporocysts","Cercariae","2d_Somules","Juveniles")) +
  labs(y = "PC2 (22.0%)",x = "PC1 (48.2%)",title = "Principal Component Analysis \n Top 25% Variable Genes") +
  theme (panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         panel.background = element_blank(),
         legend.key = element_blank(),
         axis.line = element_line(colour = "gray")) +
  ggeasy::easy_center_title() +
  scale_alpha(guide = 'none')

var25pca_plot
```


```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/var25PCA.tiff", height = 7, width = 7, units = "in", device = 'tiff', dpi = 700)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/var25PCA.png", height = 7, width = 7, units = "in", device = 'png', dpi = 1000)
```

```{r}
write.table(variable25, file = "~/Desktop/Smansoni_rnaseq/02_ANALYSIS/variable25.txt", sep = "\t",
            row.names = TRUE, col.names = NA)
```

Which genes have the most influence on each PC axis:

```{r}
pc25_loadings <- var25_pca$rotation
pc25_loadings <- pc25_loadings %>% 
  as_tibble(rownames = "gene")

pc25_loadings
```

What are the top 10 genes with highest loading on PC1 and PC2:

```{r}
slice <- dplyr::slice
top25_genes <- pc25_loadings %>% 
  # select only the PCs we are interested in
  select(gene, PC1, PC2) %>%
  # convert to a "long" format
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  # for each PC
  group_by(PC) %>% 
  # arrange by descending order of loading
  arrange(desc(abs(loading))) %>% 
  # take the 10 top rows
  slice(1:10) %>% 
  # pull the gene column as a vector
  pull(gene)

top25_loadings <- pc25_loadings %>% 
  filter(gene %in% top_genes)
```

```{r}
write.table(top25_loadings$gene, file = "~/Desktop/Smansoni_rnaseq/02_ANALYSIS/top25_loadings_genes.txt", sep = "\t",
            row.names = FALSE, col.names = FALSE)
```

Visualise variable loadings:

```{r}
loadings25_plot <- ggplot(data = top25_loadings) +
  geom_segment(aes(x = 0, y = 0, xend = PC1, yend = PC2), 
               arrow = arrow(length = unit(0.1, "in")),
               colour = "brown") +
  geom_text(aes(x = PC1, y = PC2, label = gene),
            nudge_y = 0.005, size = 3) +
  scale_x_continuous(expand = c(0.02, 0.02))

loadings25_plot
```

```{r}
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/loadings25.tiff", height = 7, width = 7, units = "in",device = 'tiff', dpi = 700)
ggsave("~/Desktop/Smansoni_rnaseq/02_ANALYSIS/loadings25.png", height = 7, width = 7, units = "in",device = 'png', dpi = 700)
```